akka-actor/src/main/scala/akka/dispatch/SuspendDetector.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.dispatch
+
+import scala.concurrent.duration._
+
+import akka.actor.ActorSystem
+import akka.actor.ClassicActorSystemProvider
+import akka.actor.ExtendedActorSystem
+import akka.actor.Extension
+import akka.actor.ExtensionId
+import akka.actor.ExtensionIdProvider
+import akka.event.Logging
+
+object SuspendDetector extends ExtensionId[SuspendDetector] with ExtensionIdProvider {
+  override def get(system: ActorSystem): SuspendDetector = super.get(system)
+
+  override def get(system: ClassicActorSystemProvider): SuspendDetector = super.get(system)
+
+  override def lookup = SuspendDetector
+
+  override def createExtension(system: ExtendedActorSystem): SuspendDetector = new SuspendDetector(system)
+
+  /**
+   * Published to the ActorSystem's eventStream when the process suspension has been detected.
+   * Note that this message could be stale when it is received and additional check with
+   * [[SuspendDetected#wasSuspended]] is recommended to be sure that the suspension occurred recently.
+   */
+  final class SuspendDetected(suspendDetectedNanoTime: Long) {
+    def wasSuspended(since: FiniteDuration): Boolean =
+      (System.nanoTime() - suspendDetectedNanoTime <= since.toNanos)
+  }
+
+}
+
+class SuspendDetector(val system: ExtendedActorSystem) extends Extension {
+  import SuspendDetector.SuspendDetected
+
+  // FIXME config
+  private val tickInterval = 100.millis
+  private val tickDeadlineNanos = 5.seconds.toNanos // FIXME default should be > 30 seconds
+  
+  private val log = Logging(system, classOf[SuspendDetector])
+
+  @volatile private var aliveTime = System.nanoTime()
+  @volatile private var suspendDetectedTime = aliveTime - 1.day.toNanos
+
+  system.scheduler.scheduleWithFixedDelay(tickInterval, tickInterval) { () =>
```

Comment:
Wouldn't this approach (scheduling `checkTime` on the scheduler and dispatcher) yield a false positive in cases of thread starvation within the application?  Thread starvation like that is definitely https://www.youtube.com/watch?v=MpUWrl3-mc8, but the response of pre-emptively tearing things down might be out of proportion; so we have a very loose deadline.

It seems like the intent is to capture scenarios where nothing in the JVM process gets scheduled (suspended process/VM, but also sufficiently-high levels of host load or priority inversions etc.), so maybe having the check run in a thread that is nearly always sleeping would be more effective and could have a faster reaction time?  My understanding is that the kernel schedulers tend to be eager to give CPU to processes that haven't been runnable in a long time (e.g. because they're sleeping).